### ISSUE TITLE
Error ZWEL0157E: Zowe dataset prefix (zowe.setup.dataset.prefix) is not defined

### ISSUE THREAD

**Issue by tsutter-wildfire** [zowe/zowe-install-packaging#3411]
Down loaded the PSWI  version of zowe 2.2.   During this part of the install
Properties for Workflow Step 3.1. Run the zowe install.  I receive a return code of 3328.
Looks like reading the zowe.yaml file is is not finding a variable .  I've looked at this file and I think it looks ok
>> INSTALL ZOWE MVS DATA SETS                                                   
                                                                                
+ export JAVA_HOME=/usr/lpp/java/J8.0_64                                        
+ export NODE_HOME=/usr/lpp/IBM/cnj/v12r0                                       
+ export PATH=/bin:/zowe/bin                                                    
+ zwe install -c /zowe/zowe.yaml                                                
  * Output:                                                                     
    node: read_yaml 10: /zowe/bin/commands/install/index.sh 30: .: zwecli_proces
Error ZWEL0157E: Zowe dataset prefix (zowe.setup.dataset.prefix) is not defined 
===============================================================================
>> INSTALL ZOWE MVS DATA SETS

  * Output:
    node: read_yaml 10: /zowe/bin/commands/install/index.sh 30: .: zwecli_process_command 4: ./zwe 96: FSUM7351 not found
Error ZWEL0157E: Zowe dataset prefix (zowe.setup.dataset.prefix) is not defined in Zowe YAML configuration file.

1.
2.

**Expected behavior**

**Screenshots**

**Logs**
If applicable, add server logs collected at the time of your problem.

**Details**
 - Version and build number: [e.g. 0.4.4-SNAPSHOT build # 155]
 - Test environment: [e.g. zD&T, zPDT, MONOPLEX, SYSPLEX, z/OS Rev 2.2, TSS, RACF, ACF2, or any other germane detail]

**Web Browser Details (if the bug relates to Zowe 
 - OS: [e.g. macOS, Windows]
 - Browser [e.g. Chrome, Safari]
 - Version [e.g. 71.0.3578.98]

**REST API client (in case of REST API issue):**
 - Technology: [e.g. Spring Boot, Node.js, Postman]
 - OS: [e.g. Windows 10]

**Shell Environment Details (if the bug relates to CLI):**
 - Technology: [e.g. zsh, cygwin, cmd, powershell, bash]
 - OS: [e.g. MacOS 10.13, Windows 10, Ubuntu 16.04 LTS]

**Additional context**

**Comment by 1000TurquoisePogs** [comment]
did you define zowe.setup.dataset.prefix? Here's what it looks like by default https://github.com/zowe/zowe-install-packaging/blob/v2.x/master/example-zowe.yaml#L32-L42
if you defined it but are still having the error, maybe it's a syntax error? YAML is whitespace sensitive, so checking that the YAML is syntactically correct may be the solution.

**Comment by tsutter-wildfire** [comment]
Yes it was in the YAML file.  I even got rid of all lines in the created YAML, that were either comment, or blanks.
This YAML was created by the installation process, so I would have expected it to be correct.
So far I’m not impressed with Zowe

**Comment by rjchavez** [comment]
Having the same issue here.  Is there a resolution or circumvention?

**Comment by tsutter-wildfire** [comment]
Not that I’ve been able to find.  It looks like whatever software reads the zowe.yaml file is not doing it correctly, or the install  process is not creating it properly
I’ve searched for answers, but google returns so much crap it’s hard to find what you are looking for .
Duckduckgo doesn’t return enough stuff

**Comment by rjchavez** [comment]
In looking at the index.sh file that's used to run the script, it looks like a read_yaml command is used.  I don't know if YAML is a requirement for USS installation but it's not stated anywhere in the doc.

**Comment by rjchavez** [comment]
I was able to get this to work from the SMP/E installation of Zowe and not from the convenience build

**Comment by tsutter-wildfire** [comment]
What did you do to make it work?  I tried both versions with the same results

**Comment by struga0258** [comment]
I was able to generate this issue when my user didn't have enough cpu to run node, can you please check and see if you are able to run node or have enough memory under your user?

**Comment by tsutter-wildfire** [comment]
Yes I can start and run node on the id that I’m using for the zowe install

**Comment by Dino390** [comment]
Hi team,
new in Zowe installation, got the same trouble.
Node seems to run fine on my z/OS V2.5 :
 >node --version
Product registration data:
Product Name: NODE.JS FOR Z/OS
PID Name: 5655-NOE
Major Version: 16
Product Owner: IBM CORP
Feature Name: NODE.JS
v16.14.2
but when doing zwe init mvs -c /...Zowe.yalm got 
-------------------------------------------------------------------------------
>> Initialize Zowe custom data sets

Error ZWEL0157E: Zowe dataset prefix (zowe.setup.dataset.prefix) is not defined in Zowe YAML configuration file.

This is true with Zowe lvl UO02050 and also lvl UO02052.

Any help would be appreciated.

**Comment by chrismcgavic** [comment]
Including a link for the yaml: https://docs.zowe.org/stable/appendix/zowe-yaml-configuration/#configure-component-app-server

I've been working on 2.3.0 zowe for the past week and have learned that the yaml is indeed sensitive to not only white space but the order in which parms are placed.  See the link above, it shows the expected order and indentation of parms.  This drive me crazy and of course still having issues with 2.3.0.  Just opened a case.

**Comment by chrismcgavic** [comment]
Also, do not do as I did and believe that the order in the sample yaml is correct, its not.  I had to move DNAME: above where it is commented out.   I would also encourage you to set useConfigmgr: to true.  Zowe will come up quicker and it does syntax checking based on the yaml link I sent above.

**Comment by lnlyssg** [comment]
I was getting a similar issue (albeit without the node message):

```$ zwe install                                                                   
=============================================================================== 
>> INSTALL ZOWE MVS DATA SETS                                                   
                                                                                
  * Output:                                                                     
    Error: error reading input file: All collection items must start at the same
 column                                                                         
Error ZWEL0157E: Zowe dataset prefix (zowe.setup.dataset.prefix) is not defined 
in Zowe YAML configuration file.                                                
```

After passing my YAML through an online syntax checker I discovered the issue was actually further down the file where I had removed some # comment prefixes but _not_ the space before/after so everything was one character out (e.g. security: on line 66 needs to be in col 5, product: in col 7, admin: in col 9 etc.), once I corrected this I was able to run zwe install OK.

**Comment by 1000TurquoisePogs** [comment]
Yes, YAML is sensitive to whitespace and also has a notion of hierarchy so where you put for example "prefix", does matter.
For example:
```
zowe:
  setup:
    dataset:
      prefix: 'abc'
```
That's zowe.setup.dataset.prefix.


```
zowe:
  setup:
    dataset:
    prefix: 'abc'
```
That's zowe.setup.prefix

```
zowe:
  setup:
    dataset:
components:
    app-server:
      prefix: 'abc'
```
That's components.app-server.prefix.

And sure, you can have a propery named "prefix" in different places all of which could be valid, but only one of those is the specific property being asked for.

Zowe server install starting with 2.3 has a new mode that I'll call "configmgr mode" which when toggled on (opt-in in 2.3, expected to be default in 2.4) it will do more validation on your yaml files and should spot issues sooner & more accurately.
You can give it a try here https://docs.zowe.org/stable/user-guide/configmgr-using#using-zwe-with-configuration-manager
It won't solve all problems for everyone at once, but over time we'll be improving it more to give you much better clues on if your YAML is valid for Zowe, and where the invalid part is.

**Comment by lnlyssg** [comment]
I did try configmgr as according to https://docs.zowe.org/stable/appendix/zwe_server_command_reference/zwe/init/zwe-init-mvs it should work for the `zwe init` command but I'm guessing the docs are further ahead than the code as it didn't actually help:

```# zwe init mvs --security-dry-run --configmgr                               
                                                                                
------------------------------------------------------------------------------- 
>> Initialize Zowe custom data sets                                             
                                                                                
  * Output:                                                                     
    Error: error reading input file: All collection items must start at the same
 column                                                                         
Error ZWEL0157E: Zowe dataset prefix (zowe.setup.dataset.prefix) is not defined 
in Zowe YAML configuration file.

**Comment by swelton** [comment]
ZOWE 2.4, z/OS 2.5, Node.js v16 and it doesn't work. this is using a copy of the delivered example and just overtyping the dummy entries:

```
`/zwe init --config /zwe240/usr/lpp/zowe/zowe.yaml
===============================================================================
>> CONFIGURE ZOWE

-------------------------------------------------------------------------------
>> Check if need to update runtime directory, Java and/or node.js settings in Zowe YAML configuration

  * Output:
    Error: error reading input file: All collection items must start at the same column
These configurations need to be added to your YAML configuration file:

zowe:
  runtimeDirectory: "/zwe240/usr/lpp/zowe"

>> Please manually update "/zwe240/usr/lpp/zowe/zowe.yaml" before you start Zowe.

-------------------------------------------------------------------------------
>> Initialize Zowe custom data sets

  * Output:
    Error: error reading input file: All collection items must start at the same column
Error ZWEL0157E: Zowe dataset prefix (zowe.setup.dataset.prefix) is not defined in Zowe YAML configuration file.`
```

This means that either the example is incorrect or there are other problems with the installer. Just for info, my YAML has the  following

```
  setup:                                           
    # MVS data set related configurations          
    dataset:                                       
      # **COMMONLY_CUSTOMIZED**                    
      # where Zowe MVS data sets will be installed 
      prefix: ZWE240.USER                          
...
  # This value is required by ZWESLSTC to know where is Zowe runtime.
  runtimeDirectory: "/zwe240/usr/lpp/zowe"                           
...
```

**Comment by chrismcgavic** [comment]
swelton I feel your pain!  I'm sure you saw the few posts before yours but you're going to have to go over your yaml with a fine toothed comb, there must be something out of place.  Don't place too much faith in the order of the parms early in the yaml, there is a hierarchy that must be followed and hopefully the yaml example can be changed to reflect a more proper order.

**Comment by chrismcgavic** [comment]
swelton, do you have zowe: in column one ahead of setup: in column three?

**Comment by lnlyssg** [comment]
> this is using a copy of the delivered example and just overtyping the dummy entries

Did you un-comment any sections/lines too? If you've seen my post above that was where I slipped up as I had only removed the # character and not the additional space needed to get things in the right place. I would try pasting your current yaml file into a checker like https://yamlchecker.com/ which should highlight any issues.

**Comment by swelton** [comment]
It seemed to be a YAMl error. Although I used the original example and just overtyped fields and where they were commented out, I just used a space over the #, it actually needed these fields to have the # deleted and moved 2 spaces to the left! That worked but there are further errors later (RACF, VSAM, etc.) but things are slowly progressing although now it cannot import zOSMF certificates. Onward and forward, thanks for the pointers.

**Comment by 1000TurquoisePogs** [comment]
For what its worth, the reason `zwe init` still has this issue is that it doesnt yet have configmgr code. I'm working on it https://github.com/zowe/zowe-install-packaging/pull/3135 but as it's a delicate part of automation code, it may take a while longer.
I recently found that if you're a user of microsoft visual studio code, redhat has a plugin for it just named "yaml". Its supposed to have the ability to tell you if your yaml is syntactically valid, so it might be good to type up a yaml in there before proceeding, if it's already a tool in your belt.

**Comment by Corzzab** [comment]
Hi I'm new to Zowe/yaml. I found a work around for this issue. If you do a zwe install -h it give you the option below to install the dataset prefix. This help for me. 

zwe install --ds-prefix ****.****.***

**Comment by muzikovam** [comment]
The original issue was caused by the workflow incorrectly creating the yaml file. So the fixes in #3761 are only for those of you who used workflow ZWECONF.xml for the configuration.

**Comment by jp669844** [comment]
No further comments, closing. If you still have an issue with the z/OSMF PSWI/Workflow method, please open a new issue.
